pkg/controller/che/che_controller.go

```diff
@@ -761,128 +764,17 @@ func (r *ReconcileChe) Reconcile(request reconcile.Request) (reconcile.Result, e
-				}
-				if instance.Spec.Auth.IdentityProviderURL != keycloakURL {
-					instance.Spec.Auth.IdentityProviderURL = keycloakURL
-					if err := r.UpdateCheCRSpec(instance, "Keycloak URL", keycloakURL); err != nil {
-						instance, _ = r.GetCR(request)
-						return reconcile.Result{Requeue: true, RequeueAfter: time.Second * 1}, err
-					}
-				}
-			} else {
-				// create Keycloak route
-				route, err := deploy.SyncRouteToCluster(deployContext, "keycloak", "", "keycloak", 8080)
-				if !tests {
-					if route == nil {
-						logrus.Info("Waiting on route 'keycloak' to be ready")
-						if err != nil {
-							logrus.Error(err)
-						}
-
-						return reconcile.Result{RequeueAfter: time.Second * 1}, err
-					}
-
-					keycloakURL := protocol + "://" + route.Spec.Host
-					if instance.Spec.Auth.IdentityProviderURL != keycloakURL {
-						instance.Spec.Auth.IdentityProviderURL = keycloakURL
-						if err := r.UpdateCheCRSpec(instance, "Keycloak URL", keycloakURL); err != nil {
-							instance, _ = r.GetCR(request)
-							return reconcile.Result{Requeue: true, RequeueAfter: time.Second * 1}, err
-						}
-						instance.Status.KeycloakURL = keycloakURL
-						if err := r.UpdateCheCRStatus(instance, "status: Keycloak URL", keycloakURL); err != nil {
-							instance, _ = r.GetCR(request)
-							return reconcile.Result{Requeue: true, RequeueAfter: time.Second * 1}, err
-						}
-					}
-				}
-			}
-
-			deploymentStatus := deploy.SyncKeycloakDeploymentToCluster(deployContext)
-			if !tests {
-				if !deploymentStatus.Continue {
-					logrus.Info("Waiting on deployment 'keycloak' to be ready")
-					if deploymentStatus.Err != nil {
-						logrus.Error(deploymentStatus.Err)
-					}
-
-					return reconcile.Result{Requeue: deploymentStatus.Requeue}, deploymentStatus.Err
-				}
-			}
-
-			if !tests {
-				if !instance.Status.KeycloakProvisoned {
-					if err := deploy.ProvisionKeycloakResources(deployContext); err != nil {
-						logrus.Error(err)
-						return reconcile.Result{RequeueAfter: time.Second}, err
-					}
-
-					for {
-						instance.Status.KeycloakProvisoned = true
-						if err := r.UpdateCheCRStatus(instance, "status: provisioned with Keycloak", "true"); err != nil &&
-							errors.IsConflict(err) {
-							instance, _ = r.GetCR(request)
-							continue
-						}
-						break
-					}
-				}
-			}
-
-			if isOpenShift {
-				doInstallOpenShiftoAuthProvider := instance.Spec.Auth.OpenShiftoAuth
-				if doInstallOpenShiftoAuthProvider {
-					openShiftIdentityProviderStatus := instance.Status.OpenShiftoAuthProvisioned
-					if !openShiftIdentityProviderStatus {
-						if err := r.CreateIdentityProviderItems(instance, request, cheFlavor, deploy.KeycloakDeploymentName, isOpenShift4); err != nil {
-							return reconcile.Result{Requeue: true, RequeueAfter: time.Second * 5}, err
-						}
-					}
-				}
+	provisioned, err := deploy.SyncIdentityProviderToCluster(deployContext, cheHost, protocol, cheFlavor)
```

Comment:
I like this refactoring.

Reply:
Well, I more or less just moved a giant if-block into a separate file :smile: It could be improved still.